zio-http/shared/src/main/scala/zio/http/codec/internal/EncoderDecoder.scala

```diff
@@ -220,171 +222,291 @@ private[codec] object EncoderDecoder {
         inputs,
         (codec, queryParams) => {
           val query      = codec.erase
-          val isOptional = query.isOptional
-          query.queryType match {
-            case QueryType.Primitive(name, bc @ BinaryCodecWithSchema(_, schema))     =>
-              val count    = queryParams.valueCount(name)
-              val hasParam = queryParams.hasQueryParam(name)
-              if (!hasParam && isOptional) None
-              else if (!hasParam) throw HttpCodecError.MissingQueryParam(name)
-              else if (count != 1) throw HttpCodecError.InvalidQueryParamCount(name, 1, count)
-              else {
-                val decoded          = bc
-                  .codec(config)
-                  .decode(
-                    Chunk.fromArray(queryParams.unsafeQueryParam(name).getBytes(Charsets.Utf8)),
-                  ) match {
+          val optional   = query.isOptionalSchema
+          val hasDefault = query.codec.defaultValue != null && query.isOptional
+          val default    = query.codec.defaultValue
+          if (codec.isPrimitive) {
+            val name     = query.nameUnsafe
+            val hasParam = queryParams.hasQueryParam(name)
+            if (
+              (!hasParam || (queryParams
+                .unsafeQueryParam(name) == "" && !emptyStringIsValue(codec.codec.schema))) && hasDefault
+            )
+              default
+            else if (!hasParam)
+              throw HttpCodecError.MissingQueryParam(name)
+            else if (queryParams.valueCount(name) != 1)
+              throw HttpCodecError.InvalidQueryParamCount(name, 1, queryParams.valueCount(name))
+            else {
+              val decoded          =
+                codec.codec.stringCodec.decode(queryParams.unsafeQueryParam(name)) match {
                   case Left(error)  => throw HttpCodecError.MalformedQueryParam(name, error)
                   case Right(value) => value
                 }
-                val validationErrors = schema.validate(decoded)(schema)
-                if (validationErrors.nonEmpty) throw HttpCodecError.InvalidEntity.wrap(validationErrors)
-                if (isOptional && decoded == None && emptyStringIsValue(schema.asInstanceOf[Schema.Optional[_]].schema))
-                  Some("")
-                else decoded
+              val validationErrors = codec.codec.erasedSchema.validate(decoded)(codec.codec.erasedSchema)
+              if (validationErrors.nonEmpty) throw HttpCodecError.InvalidEntity.wrap(validationErrors)
+              else decoded
+            }
+
+          } else if (codec.isCollection) {
```

Comment:
Definitely need to adopt the above pattern to deal with this. Some pre-analysis with pre-collection of required information, packaged into a `sealed trait`.